endpoint.go

```diff
@@ -576,39 +577,68 @@ func doUpdateHostsFile(n *network, sb *sandbox) bool {
 }
 
 func (ep *endpoint) rename(name string) error {
-	var err error
+	var (
+		err      error
+		netWatch *netWatch
+		ok       bool
+	)
+
 	n := ep.getNetwork()
 	if n == nil {
 		return fmt.Errorf("network not connected for ep %q", ep.name)
 	}
 
-	n.getController().Lock()
-	netWatch, ok := n.getController().nmap[n.ID()]
-	n.getController().Unlock()
+	c := n.getController()
 
-	if !ok {
-		return fmt.Errorf("watch null for network %q", n.Name())
+	if c.isAgent() {
+		if err = ep.deleteServiceInfoFromCluster(); err != nil {
+			return types.InternalErrorf("Could not delete service state for endpoint %s from cluster on rename: %v", ep.Name(), err)
+		}
+		// Add enough delay to avoid the above delete event be suppressed
+		// in favor of the subsequent add event while queued to be gossiped
+		// to the other nodes in the cluster
+		time.Sleep(50 * time.Millisecond)
```

Comment:
good point, the msgs will be sent in order, but could be delivered in reverse order.
In that case, we may still need to keep this sleep.
At the end it is for a very specific rename case only.